lxd/instances_post.go

```diff
@@ -404,6 +405,170 @@ func createFromMigration(s *state.State, r *http.Request, projectName string, pr
 	return operations.OperationResponse(op)
 }
 
+// createFromConversion uploads the image to the server's backup directory, converts it into the
+// the raw format, and imports the LXD instance using the converted image.
+func createFromConversion(s *state.State, r *http.Request, projectName string, profiles []api.Profile, req *api.InstancesPost) response.Response {
+	if s.DB.Cluster.LocalNodeIsEvacuated() {
+		return response.Forbidden(fmt.Errorf("Cluster member is evacuated"))
+	}
+
+	// Validate migration mode.
+	if req.Source.Mode != "push" {
+		return response.NotImplemented(fmt.Errorf("Mode %q not implemented", req.Source.Mode))
+	}
+
+	// Parse the architecture name
+	architecture, err := osarch.ArchitectureId(req.Architecture)
+	if err != nil {
+		return response.BadRequest(err)
+	}
+
+	dbType, err := instancetype.New(string(req.Type))
+	if err != nil {
+		return response.BadRequest(err)
+	}
+
+	// Only virtual machines support additional conversion options.
+	if dbType != instancetype.VM && len(req.Source.ConversionOptions) > 0 {
+		return response.BadRequest(fmt.Errorf("Conversion options can only be used with virtual machines. Instance type %q does not support conversion options", req.Type))
+	}
+
+	// Validate conversion options.
+	for _, opt := range req.Source.ConversionOptions {
+		if !slices.Contains([]string{"format"}, opt) {
+			return response.BadRequest(fmt.Errorf("Invalid conversion option %q", opt))
+		}
+	}
+
```

Comment:
There's some duplicated sections in here that could do with being moved into smaller functions and called from both code paths.